### ISSUE TITLE
Histogram Binning 1# Implementation

### ISSUE THREAD

**Issue by davyddzhalaiansamani** [Cortys/SEP-Probly-WS2526#198]
Here are 2 potential Histogram Binning implementation as a function that takes predicted values and binary labels as paramethers and return the calibrated values in order, it defines the Bins based on the actual values of the predictions: 

```python
def histogramBinning (predictions, binaryLabels): 
    
    minPre = maxPre = predictions[0]

    minPre = min(predictions)
    maxPre = max(predictions)

    k = 10 #number of bins
    binWidth = (maxPre - minPre)/k

    binSize = [0]*k
    binPositives = [0]*k

    for prediction, label in zip(predictions, binaryLabels) : 
        binId = int((prediction - minPre)/binWidth)

        if binId == k : binId = k-1
        
        binSize[binId] += 1
        binPositives[binId] += label 

    binPro = [0]*k

    for i in range(k) : 
        if binSize[i] >= 1 : 
            binPro[i] = binPositives[i]/binSize[i]
        else : binPro[i] = 0.0

    calibratedPro = []

    for prediction in predictions : 
        binId = int((prediction - minPre)/binWidth)
        
        if binId == k : binId = k-1

        calibratedPro.append(binPro[binId])

    return calibratedPro
```

The second implementation covers the case, where the confidence values are not already in the set [0, 1]

```python
from __future__ import annotations

import torch
from torch import nn

def histogram_binning_maths(predictions, truth_labels):
    sigmoid = nn.Sigmoid()
    pred_tensor = torch.tensor(predictions, dtype=torch.float32)
    calibrated_preds = sigmoid(pred_tensor)
    calibrated_preds_list = calibrated_preds.tolist()

    bins = [round(0.0 + i * 0.1, 2) for i in range(11)]
    list_of_dicts = [{0: 0, 1: 0} for _ in range(10)]
    list_of_lists = [[] for _ in range(10)]

    
    for i in range(len(calibrated_preds_list)):
        for j in range(len(bins) - 1):
            if calibrated_preds_list[i] <= bins[j + 1] or calibrated_preds_list[i] == 0:
                list_of_lists[j].append(calibrated_preds_list[i])
                list_of_dicts[j][truth_labels[i]] += 1
                break
    
    bin_thetas = []

    for i in range(len(list_of_dicts)):
        if len(list_of_lists[i]) == 0:
            bin_thetas.append(0.0)
        else:
            bin_thetas.append(list_of_dicts[i][1] / len(list_of_lists[i]) ) 

```

These functions represent first drafts of implementing histogram binning and will need to polished more to better fit  the probly repo
